pkg/azuredisk/azuredisk_v2.go

```diff
@@ -180,10 +183,11 @@ func (d *DriverV2) Run(endpoint, kubeconfig string, disableAVSetNodes, testingMo
 
 	// d.crdProvisioner is set by NewFakeDriver for unit tests.
 	if d.crdProvisioner == nil {
-		d.crdProvisioner, err = provisioner.NewCrdProvisioner(d.kubeConfig, d.config.ObjectNamespace)
-		if err != nil {
-			klog.Fatalf("Failed to get crd provisioner. Error: %v", err)
-		}
+		kubeInformerFactory := informers.NewSharedInformerFactory(d.kubeClient, consts.DefaultInformerResync)
+		azInformerFactory := azdiskinformers.NewSharedInformerFactory(d.azdiskClient, consts.DefaultInformerResync)
+		d.conditionWatcher = watcher.NewConditionWatcher(azInformerFactory, d.config.ObjectNamespace)
+		d.crdProvisioner = provisioner.NewCrdProvisioner(d.azdiskClient, d.config.ObjectNamespace, d.conditionWatcher, kubeInformerFactory, azInformerFactory)
+		d.crdProvisioner.AddPredicatesAndStartInformers(context.Background(), kubeInformerFactory, azInformerFactory, d.conditionWatcher)
```

Comment:
Architecturally, it's not a great idea for one object to rely on the side-effect of initializing another. Here, `ConditionWatcher` relies on `CrdProvisioner` to register its handlers and synchronize the informers. It makes sense to create the informers to be passed to `NewConditionWatcher` and `NewCrdProvisioner`, each adding its respective event handlers as necessary. Then have a global function that handles starting and synchronizing the informers, e.g.

```go
func StartAndWaitForCacheSync(ctx context.Context, informers... cache.SharedInformer)
```

This function can be called after all of the informers, `ConditionWatcher` and `CrdProvisioner` have been created.

Reply:
I found SharedInformerFactory could be from different packages (but not cache), it seems I can't make this function that global. Given that, do we still want to separate `WaitForCacheSync` into a function?

Reply:
Ah, yes you are right. However, this could be resolved using the following interface definition instead of `cache.SharedInformerFactory`:

```go
type InformerFactory interface {
        Start(stopCh <-chan struct{})
	Informer() cache.SharedIndexInformer
}
```

This should allow arbitrary `SharedInformerFactory` types to be passed to the method.

Reply:
Passing the `HasSynced` method to `cache.WaitForCacheSync` is a little complicated due to the indirection through the `Informer` method. You will need to iterate over the `informers` variadic argument to build an array of `cache.InformerSynced` functions which then gets passed to `cache.WaitForCacheSync`, e.g.

```go
informerSynced := make([]InformerSynced, 0)
for _, factory := range informers {
    informerSynced = append(informerSynced, factory.Informer().HasSynced)
}

cache.WaitForCacheSync(ctx, informers...)
```